sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala

```diff
@@ -406,7 +412,7 @@ abstract class JdbcDialect extends Serializable with Logging {
 
     override def visitCast(expr: String, exprDataType: DataType, dataType: DataType): String = {
       val databaseTypeDefinition =
-        getJDBCType(dataType).map(_.databaseTypeDefinition).getOrElse(dataType.typeName)
+        getJDBCType(dataType).map(_.castAsType()).getOrElse(dataType.typeName)
```

Comment:
Please do not change the behavior here. You should override the `visitCast` in `MySQLSQLBuilder`.

Reply:
Yes, we can do that. I was wondering there might be other databases use different type in cast, and they will have to duplicate the same logic somewhere.

Anyway, let's make MySQL connector work first and think about this when we have other connectors having similar semantics.

Reply:
It's implemented in MySQLDialect now. PTAL, thanks.